src/types/date.rs

```diff
@@ -0,0 +1,119 @@
+use std;
+use std::fmt;
+use std::fmt::Debug;
+use std::error::Error;
+use neon_runtime;
+use neon_runtime::raw;
+use context::{Context};
+use context::internal::Env;
+use result::{JsResult, JsResultExt};
+use object::{Object};
+use handle::{Handle, Managed};
+use super::{Value, ValueInternal};
+
+/// A JavaScript Date object.
+#[repr(C)]
+#[derive(Clone, Copy, Debug)]
+pub struct JsDate(raw::Local);
+
+impl Value for JsDate { }
+
+impl Managed for JsDate {
+    fn to_raw(self) -> raw::Local { self.0 }
+
+    fn from_raw(_: Env, h: raw::Local) -> Self { JsDate(h) }
+}
+
+#[derive(PartialEq, PartialOrd, Clone, Debug)]
+pub struct DateError(DateErrorKind);
+
+impl DateError {
+    pub fn kind(&self) -> DateErrorKind {
+        self.0
+    }
+}
+
+impl fmt::Display for DateError {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        write!(fmt, "{}", self.0.as_str())
+    }
+}
+
+impl Error for DateError {}
+
+#[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy, Debug)]
+pub enum DateErrorKind {
+    Overflow,
+    Underflow,
+}
+
+impl DateErrorKind {
+    fn as_str(&self) -> &'static str {
+        match *self {
+            DateErrorKind::Overflow => "Uncaught RangeError: Date overflow",
+            DateErrorKind::Underflow => "Uncaught RangeError: Date underflow",
+        }
+    }
+}
+
+impl<'a> JsResultExt<'a, JsDate> for Result<Handle<'a, JsDate>, DateError> {
```

Comment:
This can be more general. We don't use the `Ok` type, so we should implement it for all `T` in case we want to use it in other places.

Unfortunately, `JsResultExt` unnecessarily requires `T: Value`. I would like to fix that in a future breaking change.

Also, FYI, the [`Result::or_else`](https://doc.rust-lang.org/std/result/enum.Result.html#method.or_else) combinator performs the same thing as this `match`:

```rust
impl<'a, T: Value> JsResultExt<'a, T> for Result<Handle<'a, T>, DateError> {
    fn or_throw<'b, C: Context<'b>>(self, cx: &mut C) -> JsResult<'a, T> {
        self.or_else(|e| cx.throw_range_error(e.0.as_str()))
    }
}
```

Lastly, this includes a micro optimization of getting the `&'static str` directly from the error kind. This eliminates allocating a `String`; however, it's likely an optimization pass would remove the allocation anyway because it's trivial to see it isn't mutated.